generator/abstractmetabuilder.cpp

```diff
@@ -2480,55 +2480,100 @@ void AbstractMetaBuilder::dumpLog()
 
 AbstractMetaClassList AbstractMetaBuilder::classesTopologicalSorted() const
 {
-    AbstractMetaClassList res;
-
-    AbstractMetaClassList classes = m_meta_classes;
-    classes.sort();
-
-    QSet<AbstractMetaClass*> noDependency;
-    QHash<AbstractMetaClass*, QSet<AbstractMetaClass* >* > hash;
-    for (AbstractMetaClass *cls :  classes) {
-        QSet<AbstractMetaClass* > *depends = new QSet<AbstractMetaClass* >();
-
+    /* This function is the standard topological of a Directed Acyclic Graph
+     * (a DAG).  It outputs a partially ordered list of the nodes in the graph
+     * such that a node is output after all of its children.
+     *
+     * In the previous implementation it accounted for around 68-69% of the
+     * entire run time of pythonqt_generator.  The implementation also leaked
+     * memory and modified a QHash while iterating over it with a QHashIterator.
+     *
+     * This version is somewhat like the python equivalent using list
+     * comprehensions.
+     */
+
+    /* Build a hash list of QSet<class> for each class.  'class' is represented
+     * by a pointer to the AbstractMetaClass from m_meta_classes.
+     */
+    ReportHandler::debugSparse(QString("TSORT: %1 meta classes")
+            .arg(m_meta_classes.count()));
+    QHash<AbstractMetaClass*, QSet<AbstractMetaClass*>> classes;
+    classes.reserve(m_meta_classes.count());
+
+    for (auto cls : m_meta_classes) {
```

Comment:
You have removed the sorting of `m_meta_classes` which you had introduce earlier for a more stable output. Isn't that needed anymore?

Reply:
Ah, just noticed your comment in the other PR - apparently it indeed doesn't matter.

Reply:
The 'sort' was a side-effect of calling the function; it modified the global m_meta_classes for no reason.  The function itself sorts the output into a new AbstractMetaClassList (this, after all, is the point) and the order within the output is randomized by the use of QSet.

I've already changed the sort function to use function names, not pointers, but the QSets necessarily have to use pointers (in case the function name is duplicated; should be impossible but who knows?)  So the original sort no longer works as, maybe, expected.

The QSet stuff could be fixed too, but it means converting the `handled `QSet to a QList then sorting that immediately before it is appended to `result`.  It might produce more reliable sorted output but I still don't understand where the randomized output comes from; this function is called right at the start of the "`generate`" function.

I don't think sorting the output is a good idea at this point; this change does not alter the final sort order (making it easy to check for errors in the output), sorting would add overhead and I strongly suspect it would still not change the final sort order making the operation (like the original sort) a waste.